File: 216.combination-sum-iii.java
/*
 * @lc app=leetcode id=216 lang=java
 *
 * [216] Combination Sum III
 */

// @lc code=start
class Solution {

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> combinationSum3(int k, int n) {
        if (k <= 0 || n <= 0) return res;
        backtrack(1, k, n, new ArrayList<>());
        return res;
    }

    private void backtrack(int s, int k, int n, List<Integer> tmpList) {
        if (k == 0) {
            if (n == 0) {
                res.add(new ArrayList<>(tmpList));
            }
            return;
        }
        int b = n < 9 ? n : 9;
        for (int i = s; i <= b; ++i) {
            tmpList.add(i);
            backtrack(i+1, k-1, n-i, tmpList);
            tmpList.remove(tmpList.size()-1);
        }
    }
}
// @lc code=end

File: go/216.combination-sum-iii.go
/*
 * @lc app=leetcode id=216 lang=golang
 *
 * [216] Combination Sum III
 */

// @lc code=start
func combinationSum3(k int, n int) [][]int {
	res := [][]int{}
	board := make([]bool, 9)
	var backtrack func(int, int, int)
	backtrack = func(idx, i, acc int) {
		if i == k {
			if acc == n {
				tmp := make([]int, k)
				t := 0
				for j := 0; j < 9; j++ {
					if board[j] {
						tmp[t] = j+1
						t++
					}
				}
				res = append(res, tmp)
			}
			return
		}
		for j := idx; j < 9; j++ {
			if acc+j+1 > n {
				break
			}
			board[j] = true
			backtrack(j+1, i+1, acc+j+1)
			board[j] = false
		}
	}

	backtrack(0, 0, 0)
	return res
}
// @lc code=end

